File: java/Perceptron.java
import java.util.List;
import java.util.ArrayList;

public class Perceptron {

	private int inputNumber;
	private List<Double> weightArray;
	private double activationThreshold;
	private double bias;
	private double biasWeight;


	public Perceptron(int inputNumber, double activationThreshold) {
		this.inputNumber = inputNumber;
		this.activationThreshold = activationThreshold;
		this.weightArray = new ArrayList<Double>();

		for (int i = 0; i < inputNumber; i++) {
			this.weightArray.add(0.0);
		}

		this.bias = 1;
		this.biasWeight = 0;
	}


	public int getInputNumber() {
		return this.inputNumber;
	}


	public double getActivationThreshold() {
		return this.activationThreshold;
	}


	public List<Double> getWeightArray() {
		return this.weightArray;
	}


	public int run(List<Double> inputArray) {

		// Input check
		if (inputArray.size() != this.inputNumber) { return -1; }

		double sum = this.bias * this.biasWeight;

		for (int i = 0; i < inputArray.size(); i++) {
			sum += inputArray.get(i) * this.weightArray.get(i);
		}

		return activationFunction(sum);
	}


	public void train(List<Double> inputArray, int expectedOutput) {

		int output = run(inputArray);

		if (expectedOutput != output) {
			this.biasWeight += (expectedOutput - output) * this.bias;

			for (int i = 0; i < inputArray.size(); i++) {
				this.weightArray.set(i, this.weightArray.get(i) + ((expectedOutput - output) * inputArray.get(i)));
			}
		}
	}


	private int activationFunction(double value) {
		if (value <= this.activationThreshold) {
			return 0;
		} else {
			return 1;
		}
	}
}

File: cpp/perceptron.cpp
#include "perceptron.h"

Perceptron::Perceptron(int input_number, double activation_threshold)
{
	this->input_number = input_number;
	this->activation_threshold = activation_threshold;

	vector<double> temp(input_number, 0.0);
	this->weight_array = temp;

	this->bias = 1.0;
	this->bias_weight = 0.0;
}


int Perceptron::get_input_number()
{
	return this->input_number;
}


double Perceptron::get_activation_threshold()
{
	return this->activation_threshold;
}


vector<double> Perceptron::get_weight_array()
{
	return this->weight_array;
}


int Perceptron::run(vector<double> input_array)
{
	// input check
	if (input_array.size() != this->input_number) { return -1; }

	double sum = this->bias + this->bias_weight;
	for (unsigned int i = 0; i < input_array.size(); i++)
	{
		sum += (input_array[i] * this->weight_array[i]);
	}

	return activation_function(sum);
}


void Perceptron::train(vector<double> input_array, int expected_output)
{
	int output = run(input_array);

	if (expected_output != output)
	{
		this->bias_weight = this->bias_weight + ((expected_output - output) * this->bias);
		
		for (unsigned int i = 0; i < input_array.size(); i++)
		{
			this->weight_array[i] += ((expected_output - output) * input_array[i]);
		}
	}
}

int Perceptron::activation_function(double value)
{
	if (value <= this->activation_threshold)
	{
		return 0;
	}
	else
	{
		return 1;
	}
}